src/librustc_mir/borrow_check/nll/constraints/graph.rs

```diff
@@ -196,3 +206,59 @@ impl<'s, 'graph, D: ConstraintGraphDirecton> graph::GraphSuccessors<'graph> for
     type Item = RegionVid;
     type Iter = Successors<'graph, D>;
 }
+
+/// An "augmented region graph" builds on a normal region graph and
+/// adds extra edges for `'static`. In particular, in a normal region
+/// graph, a constraint `'a: 'b` becomes an edge `'a -> 'b` -- since
+/// we know that `'static: 'x` for all `'x`, then we can add these
+/// edges explicitly. This results in collapsing a lot more regions
+/// into the same SCC as `'static`, yielding a perf win when
+/// type-checking large constants or other things with lots of
+/// `'static` regions.
+crate struct AugmentedRegionGraph<'s> {
```

Comment:
As I [wrote on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/122657-wg-nll/subject/.2353178.20optimize.20SCC.20computation/near/132715019), I think we should try a slightly different approach now:

So right now we have these various layers. There is the "constraint graph", which yields up constraint indices. Then you can combine that with the constraint *set* to get the region graph. 

One annoying thing here is that, because we yield up constraint **indices**, if we want to make fake constraints, say, that connect `'static` to other lifetimes, then we have to add them into the vector.

Maybe instead the constraint graph iterator should take a `&'iter ConstraintSet`. Then the iterator can yield up `OutlivesConstraint` structs **by value**.  This way, we can generate `OutlivesConstraint { sup: static, sub: other_region, locations: Locations::All }` for those implied `'static` constraints.

We can then do that **always**, in the base layer of the graph. 

Then we can remove the "augmented-region-graph" concept.

Moreover, this might help with the other regressions we were seeing, since *maybe* we can tweak the `locations` value and help our "graph walker" to identify interesting points again (maybe -- I might rather take a different approach).